### PR TITLE
Fix Issue 18212 - Usage of cfloat,cdouble,cfloat,ifloat,idouble,ireal shouldn't trigger an error in deprecated code

### PR DESCRIPTION
**Pull request by wilzbach**
The following still triggers a deprecation warning:

```d
deprecated creal foo() { return 2 + 2i; }
```

In other words a `creal` in a FuncDeclaration doesn't check its function yet.
However, this is good enough to push the complex transition at Phobos as `auto`
works just fine.

See also: https://github.com/dlang/dmd/pull/7640

### CODE REVIEW

File: src/dmd/dsymbol.d [hunk 1]
```diff
@@ -2146,3 +2135,31 @@ extern (C++) final class DsymbolTable : RootObject
...
+
+/**********************************
+* Checks whether the current scope (or any of its parents) is deprecated.
+*
+* Params:
+*     sc = scope to check for
+*
+* Returns: `true` if any parent scope is deprecated, `false` otherwise`
+*/
+final bool isInDeprecatedScope(Scope* sc)
```
Comment (JinShil) [inline src/dmd/dsymbol.d]: In general I prefer not to split things out into functions unless they have more than one caller, or the caller is excessively large.  Also, in general, I think this refactoring should be in a separate PR as it is not essential to the bug-fix at hand, and should be argued on its own individual merits.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/dsymbol.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/dsymbol.d]: Another problem with this method is it's modifying the public interface of the class, yet there is no public caller or potential caller.  If it is decided to keep this, it should probably be in its own private static function within the module.

Comment (wilzbach) [reply src/dmd/dsymbol.d]: There are two callers: `checkComplexTransition` and `checkDeprecated` - it just felt wrong to duplicate code that has a certain functionality.

Comment (JinShil) [reply src/dmd/dsymbol.d]: Gah! Sorry.  Totally missed that.  You're right on here.

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -3010,15 +3010,18 @@ extern (C++) abstract class Type : RootObject
      * Params:
      *  loc = The source location.
      */
-    final void checkComplexTransition(Loc loc)
+    final bool checkComplexTransition(Loc loc, Scope* sc)
```
Comment (JinShil) [inline src/dmd/mtype.d]: C++ interface change.  See mtype.h.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/dsymbol.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/dsymbol.d]: I believe this also needs to be reflected in the C++ interface dsymbo.h

**Comment by wilzbach** [comment]
> I believe this also needs to be reflected in the C++ interface dsymbo.h

I thought we only need to add methods to the header files that are used by the LDC or GDC?
BTW I wasn't really sure where the right place to put is. After thinking a bit about it I moved it in `Scope` at `dscope.d`.

File: src/dmd/dscope.d [hunk 3]
```diff
@@ -807,4 +807,28 @@ struct Scope
         else
             return STRUCTALIGN_DEFAULT;
     }
+
+    /**********************************
+    * Checks whether the current scope (or any of its parents) is deprecated.
+    *
+    * Returns: `true` if this or any parent scope is deprecated, `false` otherwise`
+    */
+    extern(D) bool isDeprecated()
```
Comment (JinShil) [inline src/dmd/dscope.d]: Isn't `extern(D)` the default?  I think that is redundant.

**Review by JinShil — DISMISSED** [review]

**Comment by JinShil** [comment]
>After thinking a bit about it I moved it in Scope at dscope.d.

Seems logical.

File: src/dmd/mtype.d (hunk 2 above)
Comment (wilzbach) [reply src/dmd/mtype.d]: (done)

**Review by ibuclaw — APPROVED** [review]
Ok.

**Review by JinShil — APPROVED** [review]

**Comment by wilzbach** [comment]
Somehow DAutoTest is under heavy stress recently and it hasn't picked up this PR :/
@CyberShadow is there anything easy that could be done to speed up DAutoTest? With our transition to SemaphoreCI, DAutoTest is now the slowest CI.
Cleaning up the Makefile hell at dlang.org will probably take a couple of weeks.
